### PR TITLE
prevent mocks/stubs leaking from one test into another

### LINKED ISSUE #440: Consider preventing mock objects leaking from one test into another
c.f. RSpec's ["Doubles cannot be reused in another example" scenario][1].

It might be helpful to trace the use of [this code that builds a `ExpiredTestDoubleError`][2].

[1]: https://relishapp.com/rspec/rspec-mocks/v/3-9/docs/basics/scope#doubles-cannot-be-reused-in-another-example
[2]: https://github.com/rspec/rspec-mocks/blob/fa3b6a06b63d785a594723befc34edf9eed8c0f5/lib/rspec/mocks/error_generator.rb#L123-L129

### PR DESCRIPTION
**Pull request by nitishr**
Resolves #440

### CODE REVIEW

File: lib/mocha/mock.rb [hunk 1]
```diff
@@ -370,5 +377,12 @@ def ensure_method_not_already_defined(method_name)
     def any_expectations?
       @expectations.any?
     end
+
+    private
+
+    def raise_expired_error
+      raise StubbingError, "#{mocha_inspect} was originally created in one test but has leaked into another test. "\
+        'This can lead flaky tests. Ensure that your tests correctly clean up their state after themselves.'
```
[the remark this answers is not in the document]
Comment (nitishr) [reply lib/mocha/mock.rb]: Fixed

File: lib/mocha/mock.rb [hunk 2]
```diff
@@ -370,5 +377,12 @@ def ensure_method_not_already_defined(method_name)
     def any_expectations?
       @expectations.any?
     end
+
+    private
+
+    def raise_expired_error
+      raise StubbingError, "#{mocha_inspect} was originally created in one test but has leaked into another test. "\
```
[the remark this answers is not in the document]
Comment (nitishr) [reply lib/mocha/mock.rb]: Done

File: lib/mocha/mock.rb [hunk 3]
```diff
@@ -307,6 +308,7 @@ def all_expectations
     # @private
     # rubocop:disable Style/MethodMissingSuper,Metrics/CyclomaticComplexity,Metrics/PerceivedComplexity
     def method_missing(symbol, *arguments, &block)
+      raise_expired_error if @expired
```
[the remark this answers is not in the document]
Comment (nitishr) [reply lib/mocha/mock.rb]: Done. Further refactored to reduce cyclomatic & perceived complexity. No longer need to disable the two cops related to those.

**Comment by floehopper** [comment]
@nitishr Thanks again for this. I've made some minor changes to method names, warning text and commit messages and manually merged the changes in 6296fd5975f1eb9c8986e8f481f5eefc3a3a4bb6.

**Pull request closed without merging**